File: php/hardcoded_secrets/third_party_keys.php
<?php
/**
 * Shopist - Third-Party API Keys
 * DEMO FILE: Intentionally vulnerable for Datadog Code Security SAST demo.
 * DO NOT USE IN PRODUCTION.
 */

// VULN 1: SendGrid API key hardcoded as a constant (Hardcoded Secret)
define('SENDGRID_API_KEY', 'SG.ngeVJkiDSqCmhybye4kMEg.xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx');

function sendTransactionalEmail($to, $subject, $body) {
    $email = new \SendGrid\Mail\Mail();
    $email->setFrom("orders@shopist.io", "Shopist");
    $email->setSubject($subject);
    $email->addTo($to);
    $email->addContent("text/html", $body);

    $sendgrid = new \SendGrid(SENDGRID_API_KEY);
    return $sendgrid->send($email);
}

// VULN 2: Google Maps API key hardcoded in the geocoding request (Hardcoded Secret)
$googleMapsApiKey = "AIzaSyD-9tSrke72PouQMnMX-a7eZSW0jkFMBWY";

function geocodeAddress($address) {
    global $googleMapsApiKey;
    $encoded = urlencode($address);
    $url     = "https://maps.googleapis.com/maps/api/geocode/json?address={$encoded}&key={$googleMapsApiKey}";
    $resp    = file_get_contents($url);
    return json_decode($resp, true);
}

function getShippingZone($address) {
    $geo = geocodeAddress($address);
    return $geo['results'][0]['geometry']['location'] ?? null;
}

// VULN 3: Twilio Account SID and Auth Token hardcoded for SMS notifications (Hardcoded Secret)
$twilioAccountSid = "AC1234567890abcdef1234567890abcdef";
$twilioAuthToken  = "a1b2c3d4e5f67890a1b2c3d4e5f67890";

function sendShippingNotification($toPhone, $orderId, $trackingNumber) {
    global $twilioAccountSid, $twilioAuthToken;

    $client  = new Twilio\Rest\Client($twilioAccountSid, $twilioAuthToken);
    $message = $client->messages->create(
        $toPhone,
        [
            'from' => '+15005550006',
            'body' => "Your Shopist order #{$orderId} has shipped! Tracking: {$trackingNumber}",
        ]
    );
    return $message->sid;
}

File: go/hardcoded_secrets/third_party_keys.go
package hardcodedsecrets

import (
	"fmt"
	"net/http"
)

// VULN 1: Hardcoded SendGrid API key - promotional email campaigns
func SendPromoEmail(toEmail, promoCode string) error {
	sendgridAPIKey := "SG.aBcDeFgHiJkLmNoPqRsTuV.WxYz1234567890abcdefghijklmnopqrstuvwxyz12"
	req, err := http.NewRequest("POST", "https://api.sendgrid.com/v3/mail/send", nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+sendgridAPIKey)
	req.Header.Set("Content-Type", "application/json")
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return nil
}

// VULN 2: Hardcoded Google Maps API key - store locator feature
func GetStoreLocation(address string) (string, error) {
	googleMapsAPIKey := "AIzaSyDdI0hiBtXV-3Zr5elSHVQG8mEVRnS_kXc"
	url := fmt.Sprintf(
		"https://maps.googleapis.com/maps/api/geocode/json?address=%s&key=%s",
		address, googleMapsAPIKey,
	)
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	return url, nil
}

// VULN 3: Hardcoded Twilio SID and auth token - SMS order notifications
func SendOrderSMS(toPhone, message string) error {
	twilioAccountSID := "ACa1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6"
	twilioAuthToken := "a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6"
	url := fmt.Sprintf("https://api.twilio.com/2010-04-01/Accounts/%s/Messages.json", twilioAccountSID)
	req, err := http.NewRequest("POST", url, nil)
	if err != nil {
		return err
	}
	req.SetBasicAuth(twilioAccountSID, twilioAuthToken)
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return nil
}
